crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,3214 @@
+        match self {
+            SymbolExpr::Symbol(e) => e.bind(maps),
+            SymbolExpr::Value(e) => SymbolExpr::Value(e.clone()),
+            SymbolExpr::Unary(e) => e.bind(maps),
+            SymbolExpr::Binary(e) => e.bind(maps),
+        }
+    }
+
+    /// substitute symbol node to other expression
+    pub fn subs(&self, maps: &HashMap<String, SymbolExpr>) -> SymbolExpr {
+        match self {
+            SymbolExpr::Symbol(e) => e.subs(maps),
+            SymbolExpr::Value(e) => SymbolExpr::Value(e.clone()),
+            SymbolExpr::Unary(e) => e.subs(maps),
+            SymbolExpr::Binary(e) => e.subs(maps),
+        }
+    }
+
+    /// evaluate the equation
+    /// if recursive is false, only this node will be evaluated
+    pub fn eval(&self, recurse: bool) -> Option<Value> {
+        match self {
+            SymbolExpr::Symbol(_) => None,
+            SymbolExpr::Value(e) => Some(e.clone()),
+            SymbolExpr::Unary(e) => e.eval(recurse),
+            SymbolExpr::Binary(e) => e.eval(recurse),
+        }
+    }
+
+    /// calculate derivative of the equantion for a symbol passed by param
+    pub fn derivative(&self, param: &SymbolExpr) -> SymbolExpr {
+        if self == param {
+            SymbolExpr::Value(Value::Real(1.0))
+        } else {
+            match self {
+                SymbolExpr::Value(_) | SymbolExpr::Symbol(_) => SymbolExpr::Value(Value::Real(0.0)),
+                SymbolExpr::Unary(e) => e.derivative(param),
+                SymbolExpr::Binary(e) => e.derivative(param),
+            }
+        }
+    }
+
+    /// expand the equation
+    pub fn expand(&self) -> SymbolExpr {
+        match self {
+            SymbolExpr::Symbol(_) => self.clone(),
+            SymbolExpr::Value(_) => self.clone(),
+            SymbolExpr::Unary(e) => e.expand(),
+            SymbolExpr::Binary(e) => e.expand(),
+        }
+    }
+
+    /// sign operator
+    pub fn sign(&self) -> SymbolExpr {
+        SymbolExpr::Unary(Arc::new(Unary {
+            op: UnaryOps::Sign,
+            expr: self.clone(),
+        }))
+    }
+
+    /// return real number if equation can be evaluated
+    pub fn real(&self) -> Option<f64> {
+        match self.eval(true) {
+            Some(v) => match v {
+                Value::Real(r) => Some(r),
+                Value::Int(r) => Some(r as f64),
+                Value::Complex(c) => Some(c.re),
+            },
+            None => None,
+        }
+    }
+    /// return imaginary part of the value if equation can be evaluated as complex number
+    pub fn imag(&self) -> Option<f64> {
+        match self.eval(true) {
+            Some(v) => match v {
+                Value::Real(_) => Some(0.0),
+                Value::Int(_) => Some(0.0),
+                Value::Complex(c) => Some(c.im),
+            },
+            None => None,
+        }
+    }
+    /// return complex number if equation can be evaluated as complex
+    pub fn complex(&self) -> Option<Complex64> {
```

Comment:
I think it would be useful to add it to the python interface as well.

Reply:
https://github.com/doichanj/qiskit/blob/812e042946956f9b2bf840e49ded71896c61ea34/crates/circuit/src/py_symbol_expr.rs#L174-L185
is it not enough?